File: module-git/src/main/java/com/timecat/module/git/sgit/activities/CommitDiffActivity.java
package com.timecat.module.git.sgit.activities;

import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.webkit.JavascriptInterface;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.ProgressBar;

import androidx.appcompat.widget.ShareActionProvider;
import androidx.core.view.MenuItemCompat;

import com.timecat.element.alert.ToastUtil;
import com.timecat.page.base.utils.MenuTintUtils;
import com.timecat.page.base.view.BlurringToolbar;
import com.timecat.component.identity.Attr;
import com.timecat.module.git.R;
import com.timecat.module.git.android.activities.SheimiFragmentActivity;
import com.timecat.module.git.android.utils.CodeGuesser;
import com.timecat.module.git.android.utils.FsUtils;
import com.timecat.module.git.android.utils.Profile;
import com.timecat.module.git.sgit.database.models.Repo;
import com.timecat.module.git.sgit.repo.tasks.repo.CommitDiffTask;
import com.timecat.module.git.sgit.repo.tasks.repo.CommitDiffTask.CommitDiffResult;

import org.eclipse.jgit.diff.DiffEntry;
import org.eclipse.jgit.lib.PersonIdent;
import org.eclipse.jgit.revwalk.RevCommit;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.List;

public class CommitDiffActivity extends SheimiFragmentActivity {

    public final static String OLD_COMMIT = "old commit";
    public final static String NEW_COMMIT = "new commit";
    public final static String SHOW_DESCRIPTION = "show_description";
    private final static int REQUEST_SAVE_DIFF = 1;
    private static final String JS_INF = "CodeLoader";
    private WebView mDiffContent;
    private ProgressBar mLoading;
    private String mOldCommit;
    private String mNewCommit;
    private boolean mShowDescription;
    private Repo mRepo;
    private RevCommit mCommit;
    private List<String> mDiffStrs;
    private List<DiffEntry> mDiffEntries;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.git_activity_view_diff);
        setupActionBar();
        mDiffContent = (WebView) findViewById(R.id.fileContent);
        mLoading = (ProgressBar) findViewById(R.id.loading);

        Bundle extras = getIntent().getExtras();
        mOldCommit = extras.getString(OLD_COMMIT);
        mNewCommit = extras.getString(NEW_COMMIT);
        mShowDescription = extras.getBoolean(SHOW_DESCRIPTION);
        mRepo = (Repo) extras.getSerializable(Repo.TAG);

        String title = Repo.getCommitDisplayName(mNewCommit);
        if (mOldCommit != null)
            title += " : " + Repo.getCommitDisplayName(mOldCommit);

        setTitle(getString(R.string.git_title_activity_commit_diff) + title);
        loadFileContent();
    }

    private void loadFileContent() {
        mDiffContent.addJavascriptInterface(new CodeLoader(), JS_INF);
        mDiffContent.loadDataWithBaseURL("file:///android_asset/", HTML_TMPL,
                "text/html", "utf-8", null);
        WebSettings webSettings = mDiffContent.getSettings();
        webSettings.setJavaScriptEnabled(true);
        mDiffContent.setWebChromeClient(new WebChromeClient() {
            public void onConsoleMessage(String message, int lineNumber,
                                         String sourceID) {
                Log.d("MyApplication", message + " -- From line " + lineNumber
                        + " of " + sourceID);
            }

            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                return false;
            }
        });
        mDiffContent.setBackgroundColor(Color.TRANSPARENT);
    }

    private void setupActionBar() {
        BlurringToolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setPaddingStatusBar(this);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.git_diff_commits, menu);
        MenuItem item = menu.findItem(R.id.action_share_diff);
        ShareActionProvider shareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(item);
        final Intent shareIntent = new Intent(android.content.Intent.ACTION_SEND);
        shareIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        Uri futurePathName = Uri.fromFile(sharedDiffPathName());
        shareIntent.putExtra(Intent.EXTRA_STREAM, futurePathName);
        shareIntent.setData(futurePathName);
        shareIntent.setType("text/x-patch");

        shareActionProvider.setOnShareTargetSelectedListener(new ShareActionProvider.OnShareTargetSelectedListener() {
            public boolean onShareTargetSelected(ShareActionProvider source, Intent intent) {
                try {
                    File diff = sharedDiffPathName();
                    saveDiff(new FileOutputStream(diff));
                } catch (IOException e) {
                    ToastUtil.e_long(R.string.git_alert_file_creation_failure);
                }
                return false;
            }
        });

        shareActionProvider.setShareIntent(shareIntent);
        MenuTintUtils.tintAllIcons(menu, Attr.getIconColor(this));
        return true;
    }

    private String formatCommitInfo() {
        PersonIdent committer, author;
        committer = mCommit.getCommitterIdent();
        author = mCommit.getAuthorIdent();
        return "commit " + mNewCommit + "\n"
                + "Author:     " + author.getName() + " <" + author.getEmailAddress() + ">\n"
                + "AuthorDate: " + author.getWhen() + "\n"
                + "Commit:     " + committer.getName() + " <" + committer.getEmailAddress() + ">\n"
                + "CommitDate: " + committer.getWhen() + "\n";
    }

    private void saveDiff(OutputStream fos) throws IOException {
        /* FIXME: LOCK!!! */
        if (mCommit != null) {
            String message;
            fos.write(formatCommitInfo().getBytes());
            fos.write("\n".getBytes());
            message = mCommit.getFullMessage();
            for (String line : message.split("\n", -1)) {
                fos.write(("    " + line + "\n").getBytes());
            }
            fos.write("\n".getBytes());
        }
        for (String str : mDiffStrs) {
            fos.write(str.getBytes());
        }
    }

    private File sharedDiffPathName() {
        // Should we rather use createTempFile?
        String fname = mNewCommit;
        if (mOldCommit != null)
            fname += "_" + mOldCommit;
        return new File(FsUtils.getExternalDir("diff", true), fname + ".diff");
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_SAVE_DIFF && resultCode == RESULT_OK) {
            Uri diffUri = data.getData();
            try {
                saveDiff(getContentResolver().openOutputStream(diffUri));
            } catch (IOException e) {
                ToastUtil.e_long(R.string.git_alert_file_creation_failure);
            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int i = item.getItemId();
        if (i == android.R.id.home) {
            finish();
            return true;
        } else if (i == R.id.action_save_diff) {
            Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT)
                    .setType("text/x-patch")
                    .putExtra(Intent.EXTRA_TITLE, Repo.getCommitDisplayName(mNewCommit) + ".diff");

            startActivityForResult(intent, REQUEST_SAVE_DIFF);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private class CodeLoader {
        @JavascriptInterface
        public String getDiff(int index) {
            return mDiffStrs.get(index);
        }

        @JavascriptInterface
        public boolean haveCommitInfo() {
            return (mCommit != null);
        }

        @JavascriptInterface
        public String getCommitInfo() {
            if (mCommit == null) {
                return "";
            }

            return formatCommitInfo();
        }

        @JavascriptInterface
        public String getCommitMessage() {
            if (mCommit == null) {
                return "";
            }
            return mCommit.getFullMessage();
        }

        @JavascriptInterface
        public String getChangeType(int index) {
            DiffEntry diff = mDiffEntries.get(index);
            DiffEntry.ChangeType ct = diff.getChangeType();
            return ct.toString();
        }

        @JavascriptInterface
        public String getOldPath(int index) {
            DiffEntry diff = mDiffEntries.get(index);
            String op = diff.getOldPath();
            return op;
        }

        @JavascriptInterface
        public String getNewPath(int index) {
            DiffEntry diff = mDiffEntries.get(index);
            String np = diff.getNewPath();
            return np;
        }

        @JavascriptInterface
        public void getDiffEntries() {
            String oldCommit = mOldCommit != null ? mOldCommit : (mNewCommit + "^");
            CommitDiffTask diffTask = new CommitDiffTask(mRepo, oldCommit,
                    mNewCommit, new CommitDiffResult() {
                @Override
                public void pushResult(List<DiffEntry> diffEntries,
                                       List<String> diffStrs, RevCommit commit) {
                    mDiffEntries = diffEntries;
                    mDiffStrs = diffStrs;
                    mCommit = commit;
                    mLoading.setVisibility(View.GONE);
                    mDiffContent.loadUrl(CodeGuesser.wrapUrlScript("notifyEntriesReady();"));
                }
            }, mShowDescription);
            diffTask.executeTask();
        }

        @JavascriptInterface
        public int getDiffSize() {
            return mDiffEntries.size();
        }

        @JavascriptInterface
        public String getTheme() {
            return Profile.getCodeMirrorTheme(getApplicationContext());
        }
    }

    private static final String HTML_TMPL = "<!doctype html>"
            + "<head>"
            + " <script src=\"js/jquery.js\"></script>"
            + " <script src=\"js/highlight.pack.js\"></script>"
            + " <script src=\"js/local_commits_diff.js\"></script>"
            + " <link type=\"text/css\" rel=\"stylesheet\" href=\"css/rainbow.css\" />"
            + " <link type=\"text/css\" rel=\"stylesheet\" href=\"css/local_commits_diff.css\" />"
            + "</head><body></body>";

}

File: module-git/src/main/java/com/timecat/module/git/sgit/activities/CommitDiffActivity.kt
package com.timecat.module.git.sgit.activities

import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.widget.ProgressBar
import androidx.appcompat.widget.ShareActionProvider
import androidx.core.view.MenuItemCompat
import com.timecat.component.commonsdk.utils.override.LogUtil
import com.timecat.component.identity.Attr
import com.timecat.element.alert.ToastUtil
import com.timecat.module.git.R
import com.timecat.module.git.sgit.database.models.Repo
import com.timecat.module.git.sgit.database.models.Repo.Companion.getCommitDisplayName
import com.timecat.module.git.tasks.CommitDiffTask
import com.timecat.module.git.tasks.CommitDiffTask.CommitDiffResult
import com.timecat.module.git.utils.CodeGuesser
import com.timecat.module.git.utils.FsUtils.getExternalDir
import com.timecat.module.git.utils.Profile
import com.timecat.page.base.utils.MenuTintUtils
import com.timecat.page.base.view.BlurringToolbar
import org.eclipse.jgit.diff.DiffEntry
import org.eclipse.jgit.lib.PersonIdent
import org.eclipse.jgit.revwalk.RevCommit
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream

class CommitDiffActivity : SheimiFragmentActivity() {
    private var mDiffContent: WebView? = null
    private var mLoading: ProgressBar? = null
    private var mOldCommit: String? = null
    private var mNewCommit: String? = null
    private var mShowDescription = false
    private var mRepo: Repo? = null
    private var mCommit: RevCommit? = null
    private var mDiffStrs: List<String>? = null
    private var mDiffEntries: List<DiffEntry>? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.git_activity_view_diff)
        setupActionBar()
        mDiffContent = findViewById<View>(R.id.fileContent) as WebView
        mLoading = findViewById<View>(R.id.loading) as ProgressBar
        val extras = intent.extras
        mOldCommit = extras!!.getString(OLD_COMMIT)
        mNewCommit = extras.getString(NEW_COMMIT)
        mShowDescription = extras.getBoolean(SHOW_DESCRIPTION)
        mRepo = extras.getSerializable(Repo.TAG) as Repo?
        var title = getCommitDisplayName(mNewCommit)
        if (mOldCommit != null) title += " : " + getCommitDisplayName(mOldCommit)
        setTitle(getString(R.string.git_title_activity_commit_diff) + title)
        loadFileContent()
    }

    private fun loadFileContent() {
        mDiffContent!!.addJavascriptInterface(CodeLoader(), JS_INF)
        mDiffContent!!.loadDataWithBaseURL(
            "file:///android_asset/", HTML_TMPL,
            "text/html", "utf-8", null
        )
        val webSettings = mDiffContent!!.settings
        webSettings.javaScriptEnabled = true
        mDiffContent!!.webChromeClient = object : WebChromeClient() {
            override fun onConsoleMessage(message: String, lineNumber: Int, sourceID: String) {
                LogUtil.d("$message -- From line $lineNumber of $sourceID")
            }

            fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                return false
            }
        }
        mDiffContent!!.setBackgroundColor(Color.TRANSPARENT)
    }

    private fun setupActionBar() {
        val toolbar = findViewById<BlurringToolbar>(R.id.toolbar)
        toolbar.setPaddingStatusBar(this)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.git_diff_commits, menu)
        val item = menu.findItem(R.id.action_share_diff)
        val shareActionProvider = MenuItemCompat.getActionProvider(item) as ShareActionProvider?
        val shareIntent = Intent(Intent.ACTION_SEND)
        shareIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        val futurePathName = Uri.fromFile(sharedDiffPathName())
        shareIntent.putExtra(Intent.EXTRA_STREAM, futurePathName)
        shareIntent.data = futurePathName
        shareIntent.type = "text/x-patch"
        shareActionProvider!!.setOnShareTargetSelectedListener { source, intent ->
            try {
                val diff = sharedDiffPathName()
                saveDiff(FileOutputStream(diff))
            } catch (e: IOException) {
                ToastUtil.e_long(R.string.git_alert_file_creation_failure)
            }
            false
        }
        shareActionProvider.setShareIntent(shareIntent)
        MenuTintUtils.tintAllIcons(menu, Attr.getIconColor(this))
        return true
    }

    private fun formatCommitInfo(): String {
        val committer: PersonIdent = mCommit!!.committerIdent
        val author: PersonIdent = mCommit!!.authorIdent
        return """commit $mNewCommit
Author:     ${author.name} <${author.emailAddress}>
AuthorDate: ${author.getWhen()}
Commit:     ${committer.name} <${committer.emailAddress}>
CommitDate: ${committer.getWhen()}
"""
    }

    @Throws(IOException::class)
    private fun saveDiff(fos: OutputStream?) {
        /* FIXME: LOCK!!! */
        if (mCommit != null) {
            val message: String
            fos!!.write(formatCommitInfo().toByteArray())
            fos.write("\n".toByteArray())
            message = mCommit!!.fullMessage
            for (line in message.split("\n").dropLastWhile { it.isEmpty() }.toTypedArray()) {
                fos.write("    $line\n".toByteArray())
            }
            fos.write("\n".toByteArray())
        }
        for (str in mDiffStrs!!) {
            fos!!.write(str.toByteArray())
        }
    }

    private fun sharedDiffPathName(): File {
        // Should we rather use createTempFile?
        var fname = mNewCommit
        if (mOldCommit != null) fname += "_$mOldCommit"
        return File(getExternalDir("diff", true), "$fname.diff")
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_SAVE_DIFF && resultCode == RESULT_OK) {
            val diffUri = data!!.data
            try {
                saveDiff(contentResolver.openOutputStream(diffUri!!))
            } catch (e: IOException) {
                ToastUtil.e_long(R.string.git_alert_file_creation_failure)
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val i = item.itemId
        if (i == android.R.id.home) {
            finish()
            return true
        } else if (i == R.id.action_save_diff) {
            val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
                .setType("text/x-patch")
                .putExtra(Intent.EXTRA_TITLE, getCommitDisplayName(mNewCommit) + ".diff")
            startActivityForResult(intent, REQUEST_SAVE_DIFF)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private inner class CodeLoader {
        @JavascriptInterface
        fun getDiff(index: Int): String {
            return mDiffStrs!![index]
        }

        @JavascriptInterface
        fun haveCommitInfo(): Boolean {
            return mCommit != null
        }

        @JavascriptInterface
        fun getCommitInfo(): String = if (mCommit == null) {
            ""
        } else formatCommitInfo()

        @JavascriptInterface
        fun getCommitMessage(): String = if (mCommit == null) {
            ""
        } else mCommit!!.fullMessage

        @JavascriptInterface
        fun getChangeType(index: Int): String {
            val diff = mDiffEntries!![index]
            val ct = diff.changeType
            return ct.toString()
        }

        @JavascriptInterface
        fun getOldPath(index: Int): String {
            val diff = mDiffEntries!![index]
            return diff.oldPath
        }

        @JavascriptInterface
        fun getNewPath(index: Int): String {
            val diff = mDiffEntries!![index]
            return diff.newPath
        }

        @JavascriptInterface
        fun getDiffEntries() {
            val oldCommit = if (mOldCommit != null) mOldCommit!! else "$mNewCommit^"
            val diffTask = CommitDiffTask(
                mRepo!!, oldCommit, mNewCommit!!, object : CommitDiffResult {
                    override fun pushResult(diffEntries: List<DiffEntry>?, diffStrs: List<String>?, description: RevCommit?) {
                        mDiffEntries = diffEntries
                        mDiffStrs = diffStrs
                        mCommit = description
                        mLoading!!.visibility = View.GONE
                        mDiffContent!!.loadUrl(CodeGuesser.wrapUrlScript("notifyEntriesReady();"))
                    }
                }, mShowDescription
            )
            diffTask.executeTask()
        }

        @JavascriptInterface
        fun getDiffSize(): Int {
            return mDiffEntries!!.size
        }

        @JavascriptInterface
        fun getTheme(): String {
            return Profile.getCodeMirrorTheme(applicationContext)
        }
    }

    companion object {
        const val OLD_COMMIT = "old commit"
        const val NEW_COMMIT = "new commit"
        const val SHOW_DESCRIPTION = "show_description"
        private const val REQUEST_SAVE_DIFF = 1
        private const val JS_INF = "CodeLoader"
        private const val HTML_TMPL = ("<!doctype html>"
            + "<head>"
            + " <script src=\"js/jquery.js\"></script>"
            + " <script src=\"js/highlight.pack.js\"></script>"
            + " <script src=\"js/local_commits_diff.js\"></script>"
            + " <link type=\"text/css\" rel=\"stylesheet\" href=\"css/rainbow.css\" />"
            + " <link type=\"text/css\" rel=\"stylesheet\" href=\"css/local_commits_diff.css\" />"
            + "</head><body></body>")
    }
}